src/mlpack/core/optimizers/saga/saga_impl.hpp

```diff
@@ -0,0 +1,175 @@
+* terms of the 3-clause BSD license.  You should have received a copy of the
+* 3-clause BSD license along with mlpack.  If not, see
+* http://www.opensource.org/licenses/BSD-3-Clause for more information.
+*/
+#ifndef MLPACK_CORE_OPTIMIZERS_SAGA_SAGA_IMPL_HPP
+#define MLPACK_CORE_OPTIMIZERS_SAGA_SAGA_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "saga.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+template<typename UpdatePolicyType, typename DecayPolicyType>
+SAGAType<UpdatePolicyType, DecayPolicyType>::SAGAType(
+    const double stepSize,
+    const size_t batchSize,
+    const size_t maxIterations,
+    const double tolerance,
+    const bool shuffle,
+    const UpdatePolicyType& updatePolicy,
+    const DecayPolicyType& decayPolicy,
+    const bool resetPolicy) :
+    stepSize(stepSize),
+    batchSize(batchSize),
+    maxIterations(maxIterations),
+    tolerance(tolerance),
+    shuffle(shuffle),
+    updatePolicy(updatePolicy),
+    decayPolicy(decayPolicy),
+    resetPolicy(resetPolicy)
+{ /* Nothing to do. */ }
+
+//! Optimize the function (minimize).
+template<typename UpdatePolicyType, typename DecayPolicyType>
+template<typename DecomposableFunctionType>
+double SAGAType<UpdatePolicyType, DecayPolicyType>::Optimize(
+    DecomposableFunctionType& function,
+    arma::mat& iterate)
+{
+  // Find the number of functions to use.
+  const size_t numFunctions = function.NumFunctions();
+
+  // To keep track of where we are and how things are going.
+  double overallObjective = 0;
+  double lastObjective = DBL_MAX;
+
+  // Initialize the update policy.
+  if (resetPolicy)
+    updatePolicy.Initialize(iterate.n_rows, iterate.n_cols);
+
+  // Now iterate!
+  arma::mat gradient(iterate.n_rows, iterate.n_cols);
+  arma::mat gradient0(iterate.n_rows, iterate.n_cols);
+  arma::mat iterate0;
+
+  // Find the number of batches.
+  size_t numBatches = numFunctions / batchSize;
+  if (numFunctions % batchSize != 0)
+    ++numBatches; // Capture last few.
+
+  const size_t actualMaxIterations = (maxIterations == 0) ?
+               std::numeric_limits<size_t>::max() : maxIterations;
+
+  // initialize the gradients.
+  size_t effectiveBatchSize = std::min(batchSize, numFunctions);
+  arma::cube tableOfGradients(iterate.n_rows, iterate.n_cols, numBatches);
+  arma::mat avgGradient(iterate.n_rows, iterate.n_cols);
+
+  for (size_t f = 0, b = 0; f < numFunctions;
+    /* incrementing done manually */)
+  {
+    // Find the effective batch size (the last batch may be smaller).
+    effectiveBatchSize = std::min(batchSize, numFunctions - f);
+
+    function.Gradient(iterate, f, tableOfGradients.slice(b),
+                      effectiveBatchSize);
+
+    f += effectiveBatchSize;
+    avgGradient += tableOfGradients.slice(b);
+    b++;
+  }
+  avgGradient /= (double) numBatches; // Calculate average gradient
+
+  for (size_t i = 0; i < actualMaxIterations; ++i)
```

Comment:
In this case, is `actualMaxIterations` the number of passes over the training data?  We should be careful here, because with all other optimizers, `maxIterations` refers to the number of points that are seen---so `maxIterations / numFunctions` is the number of epochs.  This could be confusing to a user.

Reply:
Here `actualMaxIterations` is indeed number of passes over dataset. AFAIK the inner loop passes over all the `functions` or `batches` and the outer loop is for passing over all the `functions` repeatedly. Should, I name it something else? I've named it so because optimizers like `sgd` and `svrg` uses this name.

Reply:
All of the other optimizers take the `maxIterations` parameter to mean the number of points seen, so I would suggest adjusting this code here so that `maxIterations` means the same thing.

Reply:
I'm a little bit confused here. If maxiteration is the number of points seen then what is `numFunctions`. AFAIK after analysing the existing codebase `numFunctions`=`predictors.n_cols` which should be the number of points seen. Moreover, in almost every optimizer we calculate `actualMaxIteration` as `std::numeric_limits<size_t>::max()`whenever `maxiteration` is `0` by this logic it shouldn't be calculated in this way as we always had the number of points seen i.e. the number of points in dataset beforehend, we can then assign the number of points to `actualMaxiteration`.